### PR TITLE
Allow ability to specify retain key ordering in functions

### PR DESCRIPTION
**Pull request by srkukarni**
*(If this PR fixes a github issue, please add `Fixes #<xyz>`.)*

Fixes #<xyz>

*(or if this PR is one task of a github issue, please add `Master Issue: #<xyz>` to link to the master issue.)*

Master Issue: #<xyz>

### Motivation
Just like retainOrdering, sometimes applications want to retain ordering over the message keys. This pr adds support to that.
### Modifications

*Describe the modifications you've done.*

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API: (yes / no)
  - The schema: (yes / no / don't know)
  - The default values of configurations: (yes / no)
  - The wire protocol: (yes / no)
  - The rest endpoints: (yes / no)
  - The admin cli options: (yes / no)
  - Anything that affects deployment: (yes / no / don't know)

### Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)
  - If a feature is not applicable for documentation, explain why?
  - If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

### CODE REVIEW

File: pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/FunctionConfigUtils.java [hunk 1]
```diff
@@ -317,9 +319,15 @@ public static FunctionConfig convertFromDetails(FunctionDetails functionDetails)
         }
         if (functionDetails.getSource().getSubscriptionType() == Function.SubscriptionType.FAILOVER) {
             functionConfig.setRetainOrdering(true);
+            functionConfig.setRetainKeyOrdering(false);
             functionConfig.setProcessingGuarantees(FunctionConfig.ProcessingGuarantees.EFFECTIVELY_ONCE);
+        } else if (functionDetails.getSource().getSubscriptionType() == Function.SubscriptionType.KEY_SHARED) {
```
Comment (jerrypeng) [inline pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/FunctionConfigUtils.java:324]: We also need to do this in SinkConfigUtils

File: pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/FunctionConfigUtils.java [hunk 2]
```diff
@@ -317,9 +319,15 @@ public static FunctionConfig convertFromDetails(FunctionDetails functionDetails)
...
+            functionConfig.setRetainOrdering(false);
+            functionConfig.setRetainKeyOrdering(true);
+            functionConfig.setProcessingGuarantees(FunctionConfig.ProcessingGuarantees.ATLEAST_ONCE);
```
Comment (jerrypeng) [inline pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/FunctionConfigUtils.java:327]: We are we assuming ATLEAST_ONCE here?  A user can set the processing guarantee to be at most once and still set the retain ordering to be true

File: pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/FunctionConfigUtils.java [hunk 3]
```diff
@@ -142,6 +142,8 @@ public static FunctionDetails convert(FunctionConfig functionConfig, ClassLoader
         Function.SubscriptionType subType = ((functionConfig.getRetainOrdering() != null && functionConfig.getRetainOrdering())
                 || FunctionConfig.ProcessingGuarantees.EFFECTIVELY_ONCE.equals(functionConfig.getProcessingGuarantees()))
                 ? Function.SubscriptionType.FAILOVER
+                : (functionConfig.getRetainKeyOrdering() != null && functionConfig.getRetainKeyOrdering())
```
Comment (jerrypeng) [inline pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/FunctionConfigUtils.java:145]: Can we re-organize this giant conditional statement to be more readable?

if (effectively-once) {
   subType = Failover
} else if (retain order) {
   subType = failover
} else if (retain key order) {
 subType = shared_key
} else {
  subType = shared
}

File: pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/FunctionConfigUtils.java [hunk 4]
```diff
@@ -639,6 +655,16 @@ private static void doCommonChecks(FunctionConfig functionConfig) {
         if ((functionConfig.getMaxMessageRetries() == null || functionConfig.getMaxMessageRetries() < 0) && !org.apache.commons.lang3.StringUtils.isEmpty(functionConfig.getDeadLetterTopic())) {
             throw new IllegalArgumentException("Dead Letter Topic specified, however max retries is set to infinity");
         }
+        if (functionConfig.getRetainKeyOrdering() != null
+                && functionConfig.getRetainKeyOrdering()
```
Comment (jerrypeng) [inline pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/FunctionConfigUtils.java:659]: I would remove this stanza.  We should return an error if the retain order or retain key ordering is set at all

Comment (jerrypeng) [reply pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/FunctionConfigUtils.java:659]: Can you also check whether retain ordering is set?

File: pulsar-common/src/main/java/org/apache/pulsar/common/functions/FunctionConfig.java [hunk 5]
```diff
@@ -87,6 +87,7 @@
     private String logTopic;
     private ProcessingGuarantees processingGuarantees;
     private Boolean retainOrdering;
+    private Boolean retainKeyOrdering;
```
Comment (jerrypeng) [inline pulsar-common/src/main/java/org/apache/pulsar/common/functions/FunctionConfig.java:90]: Can we add some comments here for retainOrdering and retainKeyOrdering so people understand what they do and what the difference between the two is?

File: pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/FunctionConfigUtils.java [hunk 6]
```diff
@@ -317,9 +319,15 @@ public static FunctionConfig convertFromDetails(FunctionDetails functionDetails)
         }
         if (functionDetails.getSource().getSubscriptionType() == Function.SubscriptionType.FAILOVER) {
```
Comment (jerrypeng) [inline pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/FunctionConfigUtils.java:320]: This whole conditional statement doesn't see right to me.  Why are se setting the processing guarantee by looking at what the subscription is?  We already setting the processing guarantee in a previous line:

https://github.com/apache/pulsar/pull/7647/files#diff-239e167c18ea8591c85432948ab5040aL297

File: pulsar-common/src/main/java/org/apache/pulsar/common/functions/FunctionConfig.java (hunk 5 above)
Comment (jerrypeng) [reply pulsar-common/src/main/java/org/apache/pulsar/common/functions/FunctionConfig.java:90]: also need to add this for SinkConfig

File: pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/FunctionConfigUtils.java (hunk 2 above)
Comment (srkukarni) [reply pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/FunctionConfigUtils.java:327]: Yes. Even the existing code is broken in that regard. I will fix that

File: pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/FunctionConfigUtils.java (hunk 1 above)
Comment (srkukarni) [reply pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/FunctionConfigUtils.java:324]: That can be added later.

**Comment by srkukarni** [comment]
@jerrypeng addressed your comments. PTAL

**Pull request merged**